kernel/contracts/ZepCore.sol

```diff
@@ -0,0 +1,85 @@
+pragma solidity ^0.4.18;
+
+import "./ZepToken.sol";
+import "./KernelInstance.sol";
+import "./KernelRegistry.sol";
+import "zeppelin-solidity/contracts/math/SafeMath.sol";
+
+contract ZepCore {
+  using SafeMath for uint256;
+
+  event Staked(address indexed user, uint256 amount, uint256 total, bytes data);
+  event Unstaked(address indexed user, uint256 amount, uint256 total, bytes data);
+
+  uint256 public newVersionCost;
+  uint256 public developerFraction;
+
+  ZepToken private _token;
+  KernelRegistry private _registry;
+  uint256 private _totalStaked;
+  mapping(address => uint256) private _instanceVouches;
+  mapping(address => mapping (address => uint256)) private _userVouches;
+
+  function ZepCore(uint256 _newVersionCost, uint256 _developerFraction) public {
+    _token = new ZepToken();
+    _registry = new KernelRegistry();
+    developerFraction = _developerFraction;
+    newVersionCost = _newVersionCost;
+    // TODO: we need to think how we are going to manage variable costs to propose new versions
+  }
+
+  function register(KernelInstance instance) public {
+    registry().addInstance(instance);
+    token().transferFrom(msg.sender, this, newVersionCost);
+    token().burn(newVersionCost);
+    // TODO: should we build a burnFrom or sth like that?
+  }
+
+  function transferStake(KernelInstance from, KernelInstance to, uint256 amount, bytes data) public {
+    // TODO: an unstake followed by a stake call won't work
```

Comment:
Why not?

Reply:
We thought allowances would prevent it, but maybe we were wrong since the allowance should stay in place even if we return the token.. In any case it's not great that the tokens return to the staker and back to ZepCore, better to handle logic independently, trying to encapsulate repeated code.

Reply:
Regarding allowance, I'd assume that users of zepcore have set a sort-of unlimited allowance to the contract. However, it's true that "it's not great that the tokens return to the staker and back to ZepCore", as we're wasting gas unnecessarily; so it's indeed better to implement this a special case.